### PR TITLE
Change how Handlers are added in Netty Http

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Changes how handlers are added to Reactor Netty's/Netty's channel pipeline. Instead of the `doX` operations being called in `NettyAsyncHttpClient` they are now called, again, in `NettyAsyncHttpClientBuilder`. This massively reduces the number of times they are called and therefore massively reduces the number of times Reactor Netty's `HttpClient` has to duplicate configurations, as each `doX` call duplicates the underlying configurations. `doX` methods were being called in `NettyAsyncHttpClient` to enable passing contextual information to the outbound handlers and response handlers, dynamic response timeouts and progress reporting, this is now handled by adding subscriber context to the reactive HTTP request by calling `contextWrite` in `NettyAsyncHttpClient` and inspecting the `HttpClientRequest` for containing the written context.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClientBuilder.java [hunk 1]
```diff
@@ -496,4 +520,37 @@ static long getTimeoutMillis(Duration configuredTimeout, long defaultTimeout) {
...
+
+        InetSocketAddress inetSocketAddress = (InetSocketAddress) socketAddress;
+
+        return !nonProxyHostsPattern.matcher(inetSocketAddress.getHostString()).matches();
+    }
+
+    /*
+     * Request has started, add the Azure SDK ChannelAdapter.
+     */
+    private void doOnRequest(HttpClientRequest request, Connection connection, long writeTimeout, long responseTimeout,
```
Comment (srnagar) [inline sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClientBuilder.java:541]: Consider naming this `addRequestHandler` to not misinterpret this as `doOnRequest` method provided by Netty and also this describes what the method is doing.

File: sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClientBuilder.java [hunk 2]
```diff
@@ -496,4 +520,37 @@ static long getTimeoutMillis(Duration configuredTimeout, long defaultTimeout) {
...
+            AzureNettyHttpClientContext.KEY, null);
+
+        connection.addHandlerLast(AzureSdkHandler.HANDLER_NAME, new AzureSdkHandler(attr, writeTimeout, responseTimeout,
+            readTimeout));
+    }
+
+    /*
+     * Response has completed, remove the Azure SDK ChannelHandler.
+     */
+    private static void doAfterResponseSuccess(Connection connection) {
```
Comment (srnagar) [inline sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/NettyAsyncHttpClientBuilder.java:553]: Same here - consider naming this `removeHandler()`.

File: sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/AzureNettyHttpClientContext.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+// Licensed under the MIT License.
+
+package com.azure.core.http.netty.implementation;
+
+import com.azure.core.util.ProgressReporter;
+
+/**
+ * A holder for all context that may be passed through the Netty HttpClient.
+ */
+public final class AzureNettyHttpClientContext {
```
Comment (srnagar) [inline sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/AzureNettyHttpClientContext.java:11]: `responseTimeoutOverride` and `progressReporter` are already in the context. Do we need a new type to hold these two values?

File: sdk/core/azure-core-http-netty/src/test/java/com/azure/core/http/netty/NettyAsyncHttpClientProviderTests.java [hunk 4]
```diff
@@ -88,9 +88,6 @@ public void optionsWithTimeouts() {
         Integer connectTimeout = (Integer) httpClient.nettyClient.configuration().options()
             .get(ChannelOption.CONNECT_TIMEOUT_MILLIS);
         assertEquals((int) expectedTimeout, connectTimeout.intValue());
-        assertEquals(expectedTimeout, httpClient.writeTimeout);
-        assertEquals(expectedTimeout, httpClient.responseTimeout);
-        assertEquals(expectedTimeout, httpClient.readTimeout);
```
Comment (srnagar) [inline sdk/core/azure-core-http-netty/src/test/java/com/azure/core/http/netty/NettyAsyncHttpClientProviderTests.java:93]: Any reason for removing these asserts?

Comment (alzimmermsft) [reply sdk/core/azure-core-http-netty/src/test/java/com/azure/core/http/netty/NettyAsyncHttpClientProviderTests.java:93]: NettyAsyncHttpClient no longer has these properties

File: sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/AzureNettyHttpClientContext.java (hunk 3 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/AzureNettyHttpClientContext.java:11]: Reduces the number of lookups we need to do in `Context`, so if we add more configurations in the future they all can be stored in one location.

Comment (srnagar) [reply sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/AzureNettyHttpClientContext.java:11]: The number of lookups should be the same if we do the lookup in the ctor of AzureSdkHandler (and the context passed in as ctor arg). Will not require us to create a new instance of this type each time in a hot path.

**Review by srnagar — APPROVED** [review]
LGTM!

**Pull request merged**